PR : 실습 4주차 필수과제

Jokwanhee opened the pull request:
## 📌𝘐𝘴𝘴𝘶𝘦𝘴
- closed #12 

## 📎𝘞𝘰𝘳𝘬 𝘋𝘦𝘴𝘤𝘳𝘪𝘱𝘵𝘪𝘰𝘯
- Retrofit2 with MVVM 패턴+ Repository 패턴 적용
- Application 사용으로 retrofit 및 sharedPreference 초기화
- HttpLogginInterceptor 및 OkHttpClient 사용 
- 회원가입 및 로그인 API 연동
- Reqres List Users API 연동

## 📷𝘚𝘤𝘳𝘦𝘦𝘯𝘴𝘩𝘰𝘵
https://github.com/DO-SOPT-ANDROID/kwanhee-jo/assets/90740783/84eb768e-42df-4e4f-8149-1baedd71db8f

## 💬𝘛𝘰 𝘙𝘦𝘷𝘪𝘦𝘸𝘦𝘳𝘴
Retrofit 과 함께 MVVM + Repository 패턴을 적용해보았습니다. 의존성 주입을 위한 라이브러리를 사용하지 않고 하니 뭔가 코드가 복잡해보입니다. ViewModelFactory, Application 등 클래스, 인터페이스가 많아지니깐 어지럽네요..  제대로 패턴이 적용된지도 모르겠네요... 리뷰 항상 감사드립니다! (채찍질 부탁드립니다.. 지적할 부분있으면 달게 받겠습니다!)

Jokwanhee committed:
[add/#12] kotlinx.serialization and BuildConfig

- json 파싱 라이브러리 추가
- local.properties 에 base url 추가로 BuildConfig 설정

Jokwanhee committed:
[mod/#12] SharedPreferences

companion object 내부 sharedPreferences 사용으로 불필요한 getter 함수 삭제

Jokwanhee committed:
[add/#12] retrofit setting

- retrofit2
- kotlinx.serialization
- dto

Jokwanhee committed:
[add/#12] signup

- repository pattern 적용
- mvvm 적용
- application 사용 (전역 sharedPreference 초기화 및 retrofit 초기화)

Jokwanhee committed:
[add/#12] signup

commit 잘못했뜸.. ㅋㅋ
- repository pattern 적용
- mvvm 적용
- application 사용 (전역 sharedPreference 초기화 및 retrofit 초기화)

Jokwanhee committed:
[add/#12] login

로그인 api 연결

Jokwanhee committed:
[add/#12] reqres api

reqres api 연결
- LIST USERS api 사용

Jokwanhee committed:
[add/#12] regres api ui binding

List Users 데이터 바인

Jokwanhee committed:
[del/#12] homeAdapter

- 기존 homeAdapter 멀티 뷰 타입 파일 삭제
- 불필요한 코드 주석 및 삭제

kez-lab commented on app/src/main/java/org/sopt/dosopttemplate/db/remote/AuthApiHelper.kt:
```diff
@@ -0,0 +1,73 @@
+                    response: Response<Unit>
+                ) {
+                    if (response.isSuccessful && response.code() == 201) {
+                        onResponse(SignUpResp.Success(response.headers()["Location"]?.split("/")?.last().toString()))
```
isSuccessful변수가 true라는 것은 200번대라는 얘기이기 때문에 code까지 분기해서 확인할 필요는 없으실 것 같습니다~!!

kez-lab commented on app/src/main/java/org/sopt/dosopttemplate/db/remote/AuthApiHelper.kt:
```diff
@@ -0,0 +1,73 @@
+                }
+
+                override fun onFailure(call: Call<Unit>, t: Throwable) {
+                    t.printStackTrace()
```
t.printStackTrace의 경우에는 에러로그가 릴리즈 상황에서도 보이기 때문에 Log를 활용하여 erro를 찍는 것을 권장드립니다!

kez-lab commented on app/src/main/java/org/sopt/dosopttemplate/presentation/home/HomeActivity.kt:
```diff
@@ -2,66 +2,47 @@ package org.sopt.dosopttemplate.presentation.home
+        homeViewModel.getUserList()
+    }
+
+    private fun initViewModel() {
+        homeViewModel = ViewModelProvider(
+            this,
+            UserViewModelFactory(UserRepository(DoSoptApp.getUserApiHelperInstance()))
+        ).get(HomeViewModel::class.java)
     }
```
👏

kez-lab commented on app/src/main/java/org/sopt/dosopttemplate/presentation/login/LoginActivity.kt:
```diff
@@ -46,15 +51,40 @@ class LoginActivity : BaseActivity<ActivityLoginBinding>(ActivityLoginBinding::i
     override fun initView() {
         binding.soptEvId.setInputType(InputType.TYPE_CLASS_TEXT)
         binding.soptEvPwd.setInputType(InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_PASSWORD)
+        initViewModel()
         autoLogin()
+        observeData()
+    }
+
+    private fun initViewModel() {
+        loginViewModel =
+            ViewModelProvider(
+                this,
+                AuthViewModelFactory(AuthRepository(DoSoptApp.getApiHelperInstance()))
+            ).get(LoginViewModel::class.java)
```
initView함수에 initViewModel이 있는게 조금 이상하네요~!! ViewModel을 생성하는 시점이 언제여야하는지 함수의 관심사 분리와 함께 고민해보시면 좋을 것 같습니다.

kez-lab commented on app/src/main/java/org/sopt/dosopttemplate/presentation/signup/SignUpActivity.kt:
```diff
@@ -0,0 +1,77 @@
+class SignUpActivity : BaseActivity<ActivitySignUpBinding>(ActivitySignUpBinding::inflate) {
+    private lateinit var signUpViewModel: SignUpViewModel
+
+    override fun initView() {
+        initViewModel()
+        observeData()
+    }
```
Base가 initView를 강제하다보니 이렇게 되는 것 같은데 Base구조 변경을 고려하여야할 것 같습니다

kez-lab commented on app/src/main/java/org/sopt/dosopttemplate/presentation/signup/SignUpActivity.kt:
```diff
@@ -0,0 +1,77 @@
+        }
+    }
+
+    override fun initEvent() {
+        initDoSignUp()
+    }
+
+    private fun initDoSignUp() {
+        binding.btSignUp.setOnClickListener {
+            hideKeyboard(binding.root)
+            if (
+                binding.soptEvId.getEditText().length in 6..10
+                && binding.soptEvPwd.getEditText().length in 8..12
+                && binding.soptEvNickname.getEditText().isNotEmpty()
+                && binding.soptEvMbti.getEditText().toMBTI() != MBTI.ERROR
+            ) {
+                signUpViewModel.signUp(
+                    id = binding.soptEvId.getEditText(),
+                    nickname = binding.soptEvNickname.getEditText(),
+                    password = binding.soptEvPwd.getEditText()
+                )
+            }
+        }
+    }
```
함수명이 어떤 동작을 하는지 잘 모르겠어요! if문도 너무 길어서 가독성이 떨어지는 것 같습니다

kez-lab commented on app/src/main/java/org/sopt/dosopttemplate/repository/AuthRepository.kt:
```diff
@@ -0,0 +1,12 @@
+package org.sopt.dosopttemplate.repository
+
+import org.sopt.dosopttemplate.db.remote.AuthApiHelper
+import org.sopt.dosopttemplate.model.dto.RespResult
+import org.sopt.dosopttemplate.model.dto.resp.auth.SignUpResp
+
+class AuthRepository(private val authApiHelper: AuthApiHelper) {
+    suspend fun signUp(id: String, nickname: String, password: String, onResponse: (SignUpResp) -> Unit) =
+        authApiHelper.signUp(id, nickname, password, onResponse)
+    suspend fun login(id: String, password: String, auto: Boolean, onResponse: (RespResult) -> Unit) =
+        authApiHelper.login(id, password, auto, onResponse)
+}
```
authApiHelper에서 signUp과 login이 suspend 함수가 아닌데 해당 키워드를 붙여주신 이유가 있을까요??

Jokwanhee replied:
코루틴 공부한다고 해놓고 제가 안지웠네요.. 지우고 나중에 코루틴 공부하면서 적용하겠습니당!

kez-lab reviewed (changes requested):
전체적으로 주석제거 및 코멘트 확인 한번 부탁드립니다~!! 너무 고생하셨어요 👍

Jokwanhee committed:
[review] 의진 리뷰 피드백

response.isSuccessful 은 200 과 같으므로response.code 분기 처리 불필요

Jokwanhee committed:
[review] 의진 리뷰 피드백

printstacktrace 는 릴리즈 이후에도 보이기 때문에 Log 를 사용하여서 에러메시지를 확인하자

Jokwanhee committed:
[review] 의진 리뷰 피드백

BaseXXX 를 사용하면서 함수명 미스매치가 많음. Base를 상속하는 액티비티는 initView 를 무조건 불러야해서 initView 와 initEvent 삭제

Jokwanhee committed:
[review] 의진 리뷰 피드백

회원가입 함수 내부 구현 결합도 낮추기

Jokwanhee committed:
[review] 의진 리뷰 피드백

쓸데없는 suspend 키워드 삭제

Jokwanhee committed:
[review] 의진 리뷰 피드백

주석, 불필요한 코드 및 파일 삭제

Jokwanhee commented:
> 전체적으로 주석제거 및 코멘트 확인 한번 부탁드립니다~!! 너무 고생하셨어요 👍

코드 리뷰 해주셔서 감사합니다 _ _) 리뷰에 대한 피드백 적용해보았습니다 📃

kez-lab reviewed (approved):
굳굳!! 🍀

emjayMJkim commented on app/src/main/java/org/sopt/dosopttemplate/model/dto/req/LoginReq.kt:
```diff
@@ -0,0 +1,12 @@
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class LoginReq(
```
혹시 Request 끝까지 안 써주는 이유가 따로 있을까요..?? LoginReqrest 이런 식으로 쓰는 게 더 가독성이 좋을 것 같아서염..!!

Jokwanhee replied:
저는 Req = Request 인데, 다른분들이 보기에는 헷갈릴수도있겠네용.. 이번에 서버통신할 때 풀네임으로 함수네임 만들어보겠씁니다!

emjayMJkim commented on app/src/main/java/org/sopt/dosopttemplate/presentation/signup/SignUpActivity.kt:
```diff
@@ -0,0 +1,77 @@
+            ViewModelProvider(
+                this,
+                AuthViewModelFactory(AuthRepository(getApiHelperInstance()))
+            ).get(SignUpViewModel::class.java)
```
굳굳

emjayMJkim reviewed (commented):
과제 하느라 수고했어~~ 멋있다 진짜

yeonjeen commented on app/src/main/java/org/sopt/dosopttemplate/db/remote/AuthApiHelper.kt:
```diff
@@ -0,0 +1,78 @@
+                override fun onFailure(call: Call<LoginResp>, t: Throwable) {}
+            })
+    }
+}
```
callback 부분이 코드가 비슷한데 이럴 때 함수화를 시켜도 될까요?

Jokwanhee replied:
해당 부분의 반복되는 코드를 줄이기 위해서 확장함수로 사용하는 방법은 우리 안드 파트장님이 꿀팁 공유에 "우리 5인큐 하자" 라는 타이틀의 아티클을 올렸으니 한 번 확인해보면 좋을 듯 합니다!!

yeonjeen commented on app/src/main/java/org/sopt/dosopttemplate/presentation/signup/SignUpActivity.kt:
```diff
@@ -0,0 +1,77 @@
+        binding.soptEvId.getEditText().length in 6..10
+        && binding.soptEvPwd.getEditText().length in 8..12
+        && binding.soptEvNickname.getEditText().isNotEmpty()
+        && binding.soptEvMbti.getEditText().toMBTI() != MBTI.ERROR
```
이 부분 함수보면서 가독성이 정말 좋다고 느꼈어요!!

yeonjeen reviewed (commented):
코드를 보면서 가독성도 좋고 코드도 간결해서 이해가 잘 되었던 것 같아요!! 항상 많이 배우고갑니다!!
